README.md
# cmdspy [![Build Status](https://travis-ci.org/kunit/cmdspy.svg?branch=master)](https://travis-ci.org/kunit/cmdspy) [![codecov](https://codecov.io/gh/kunit/cmdspy/branch/master/graph/badge.svg)](https://codecov.io/gh/kunit/cmdspy)

引数で指定したコマンドが実行されているかを一定間隔ごとに Slack に通知します。

## 使い方

```
$ cmdspy --help
Usage: cmdspy [--version] [--help] <options> "command <arg1> <arg2>..."

Options:
  -c, --config                             path to configuration file
  -i, --interval=seconds                   report interval (default: 600)

% echo config.toml
url = "https://hooks.slack.com/services/XXXXXX/YYYYYYY/ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ"
channel = "#chanel_name"
emoji = ":emoji_name:"
interval = 1800

% cmdspy -c config.toml "/path/to/long_time_cmd -a -b -c"
```

## Slack 通知サンプル

`sh /Users/kunit/Work/test.sh` を10秒単位に通知させた結果です。

```
$ ./cmdspy -i 10 -c config.toml "sh /Users/kunit/Work/test.sh"
```
<img width="659" alt="スクリーンショット" src="https://user-images.githubusercontent.com/405750/55526937-faf49f80-56d1-11e9-9fac-c39a751733df.png">


CHANGELOG.md
# Changelog

## [v0.3.2](https://github.com/kunit/cmdspy/compare/v0.3.1...v0.3.2) (2019-04-05)

- 引数の interval が正しく処理できていなかった

## [v0.3.1](https://github.com/kunit/cmdspy/compare/v0.3.1...v0.3.1) (2019-04-05)

- 経過時間の表示がおかしかったのを修正(#1)

## [v0.3.0](https://github.com/kunit/cmdspy/compare/v0.2.0...v0.3.0) (2019-04-04)

- slack に送る内容にそのプロセスの ps の結果相当のものをつけるようにした

## [v0.2.0](https://github.com/kunit/cmdspy/compare/v0.1.0...v0.2.0) (2019-04-03)

- エラー時にメンションを飛ばせるようにした

## [v0.1.0](https://github.com/kunit/cmdspy/compare/bd4c4bae58b3...v0.1.0) (2019-04-03)

- デフォルトの監視間隔を30分(1800秒)にした
- ひとまず形にした



cmdspy.go
package cmdspy

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"time"

	slacli "github.com/monochromegane/slack-incoming-webhooks"
)

// Slack struct
type Slack struct {
	Title      string
	Message    string
	Color      string
	IconEmoji  string
	WebhookURL string
	Channel    string
	Mentions   []string
}

// Spy
func Spy(args []string, config Config, interval int) bool {
	start := time.Now()

	emoji := ":sunglasses:"
	if config.Emoji != "" {
		emoji = config.Emoji
	}

	sl := Slack{
		WebhookURL: config.Url,
		Channel:    config.Channel,
		IconEmoji:  emoji,
		Mentions:   config.Mentions,
		Title:      args[0],
		Message:    "Exec Command",
		Color:      "#5CB589",
	}

	s := strings.Split(args[0], " ")
	cmd := exec.Command(s[0], s[1:]...)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		sl.Message = err.Error()
		sl.Color = "#961D13"
		sl.Message = err.Error()
		postMessageToSlack(sl, true)
		return false
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		sl.Message = err.Error()
		sl.Color = "#961D13"
		sl.Message = err.Error()
		postMessageToSlack(sl, true)
		return false
	}

	if err := cmd.Start(); err != nil {
		sl.Message = err.Error()
		sl.Color = "#961D13"
		sl.Message = err.Error()
		postMessageToSlack(sl, true)
		return false
	}

	line, err := getPs(cmd.Process.Pid)
	if err != nil {
		sl.Message = fmt.Sprintf("%s\n%s", sl.Message, err.Error())
	} else {
		sl.Message = fmt.Sprintf("%s\n%s", sl.Message, line)
	}

	postMessageToSlack(sl, false)

	streamReader := func(scanner *bufio.Scanner, outputChan chan string, doneChan chan bool) {
		defer close(outputChan)
		defer close(doneChan)
		for scanner.Scan() {
			outputChan <- scanner.Text()
		}
		doneChan <- true
	}

	stdoutScanner := bufio.NewScanner(stdout)
	stdoutOutputChan := make(chan string)
	stdoutDoneChan := make(chan bool)
	stderrScanner := bufio.NewScanner(stderr)
	stderrOutputChan := make(chan string)
	stderrDoneChan := make(chan bool)
	go streamReader(stdoutScanner, stdoutOutputChan, stdoutDoneChan)
	go streamReader(stderrScanner, stderrOutputChan, stderrDoneChan)

	nextInterval := config.Interval
	if interval != 0 {
		nextInterval = interval
	}

	state := true
	for state {
		select {
		case <-stdoutDoneChan:
			state = false
		case line := <-stdoutOutputChan:
			fmt.Println(line)
		case line := <-stderrOutputChan:
			fmt.Println(line)
		default:
			now := time.Now()
			duration := now.Sub(start)
			if int(duration.Seconds()) >= nextInterval {
				sl.Message = fmt.Sprintf("%s から %02d:%02d:%02d 経過", start.Format("2006-01-02 15:04:05"), int(duration.Hours()), int(duration.Minutes())%60, int(duration.Seconds())%60)
				line, err := getPs(cmd.Process.Pid)
				if err != nil {
					sl.Message = fmt.Sprintf("%s\n%s", sl.Message, err.Error())
				} else {
					sl.Message = fmt.Sprintf("%s\n%s", sl.Message, line)
				}
				postMessageToSlack(sl, false)
				nextInterval += interval
			}
		}
	}

	if err := cmd.Wait(); err != nil {
		sl.Message = err.Error()
		sl.Color = "#961D13"
		postMessageToSlack(sl, true)
		return false
	} else {
		sl.Message = "Success"
		postMessageToSlack(sl, false)
	}

	return true
}

var slackReplacer = strings.NewReplacer("&", "&amp;", "<", "&lt;", ">", "&gt;")

// postMessageToSlack
func postMessageToSlack(sl Slack, withMention bool) error {
	if sl.Channel != "" && !strings.HasPrefix(sl.Channel, "#") {
		sl.Channel = "#" + sl.Channel
	}

	sl.Message = slackReplacer.Replace(sl.Message)

	cli := slacli.Client{WebhookURL: sl.WebhookURL}

	var text string
	if withMention {
		for _, mention := range sl.Mentions {
			text = text + fmt.Sprintf("%s ", mention)
		}
		if len(text) > 0 {
			text = fmt.Sprintf("%sエラーが発生しました", text)
		}
	}

	return cli.Post(&slacli.Payload{
		Username:  "cmdspy",
		IconEmoji: sl.IconEmoji,
		Channel:   sl.Channel,
		Text:      text,
		Attachments: []*slacli.Attachment{
			{
				Color:      sl.Color,
				Title:      sl.Title,
				Text:       sl.Message,
				MarkdownIn: []string{"Text"},
			},
		},
	})
}

// getPs
func getPs(pid int) (string, error) {
	ps := []string{"ps", "-p", fmt.Sprintf("%d", pid), "-o", "user,pid,%cpu,%mem,vsz,rss,tt,state,start,time"}
	c := exec.Command(ps[0], ps[1:]...)
	c.Stdin = os.Stdin
	c.Env = append(os.Environ(), "LC_ALL=POSIX")
	out, err := c.Output()
	if err != nil {
		return "", err
	}

	return fmt.Sprintf("```%s```", string(out)), nil
}


cli.go
package cmdspy

import (
	"bytes"
	"fmt"
	"github.com/BurntSushi/toml"
	"github.com/jessevdk/go-flags"
	"github.com/mgutz/ansi"
	"io"
	"reflect"
	"strings"
)

const (
	// ExitOK for exit code
	ExitOK int = 0

	// ExitErr for exit code
	ExitErr int = 1
)

const (
	OutColor = "green"

	ErrColor = "red"
)

// cli struct
type cli struct {
	env      Env
	command  string
	args     []string
	Config   string `long:"config" short:"c" description:"path to configuration file"`
	Interval int    `long:"interval" arg:"seconds" short:"i" description:"report interval (default: 600)"`
	Help     bool   `long:"help" short:"h" description:"show this help message and exit"`
	Version  bool   `long:"version" short:"v" description:"prints the version number"`
}

// Env struct
type Env struct {
	Out, Err io.Writer
	Args     []string
	Version  string
}

// Config
type Config struct {
	Url      string
	Channel  string
	Emoji    string
	Mentions []string
	Interval int
}

// RunCLI runs as cli
func RunCLI(env Env) int {
	cli := &cli{env: env, Interval: 1800}
	return cli.run()
}

// buildHelp
func (c *cli) buildHelp(names []string) []string {
	var help []string
	t := reflect.TypeOf(cli{})

	for _, name := range names {
		f, ok := t.FieldByName(name)
		if !ok {
			continue
		}

		tag := f.Tag
		if tag == "" {
			continue
		}

		var o, a string
		if a = tag.Get("arg"); a != "" {
			a = fmt.Sprintf("=%s", a)
		}
		if s := tag.Get("short"); s != "" {
			o = fmt.Sprintf("-%s, --%s%s", tag.Get("short"), tag.Get("long"), a)
		} else {
			o = fmt.Sprintf("--%s%s", tag.Get("long"), a)
		}

		desc := tag.Get("description")
		if i := strings.Index(desc, "\n"); i >= 0 {
			var buf bytes.Buffer
			buf.WriteString(desc[:i+1])
			desc = desc[i+1:]
			const indent = "                        "
			for {
				if i = strings.Index(desc, "\n"); i >= 0 {
					buf.WriteString(indent)
					buf.WriteString(desc[:i+1])
					desc = desc[i+1:]
					continue
				}
				break
			}
			if len(desc) > 0 {
				buf.WriteString(indent)
				buf.WriteString(desc)
			}
			desc = buf.String()
		}
		help = append(help, fmt.Sprintf("  %-40s %s", o, desc))
	}

	return help
}

// showHelp
func (c *cli) showHelp() {
	opts := strings.Join(c.buildHelp([]string{
		"Config",
		"Interval",
	}), "\n")

	help := `Usage: cmdspy [--version] [--help] <options> "command <arg1> <arg2>..."

Options:
%s
`
	fmt.Fprintf(c.env.Out, help, opts)
}

// run
func (c *cli) run() int {
	p := flags.NewParser(c, flags.PassDoubleDash)
	args, err := p.ParseArgs(c.env.Args)
	if err != nil || c.Help {
		c.showHelp()
		return ExitOK
	}

	if c.Version {
		fmt.Fprintf(c.env.Err, "cmdspy version %s\n", c.env.Version)
		return ExitOK
	}

	if len(c.Config) == 0 {
		fmt.Fprint(c.env.Err, ansi.Color("Error: Required config option\n\n", ErrColor))
		c.showHelp()
		return ExitErr
	}

	var config Config
	_, err = toml.DecodeFile(c.Config, &config)
	if err != nil {
		fmt.Fprint(c.env.Err, ansi.Color(fmt.Sprintf("Error: DecodeFile Failure %s\n\n%v\n\n", c.Config, err.Error()), ErrColor))
		c.showHelp()
		return ExitErr
	}

	if len(config.Url) == 0 || len(config.Channel) == 0 {
		fmt.Fprint(c.env.Err, ansi.Color("Error: Required url and channel\n\n", ErrColor))
		c.showHelp()
		return ExitErr
	}

	if len(args) == 0 {
		c.showHelp()
		return ExitErr
	}

	if !Spy(args, config, c.Interval) {
		return ExitErr
	}

	return ExitOK
}
